engine/player/covenant.cpp

```diff
@@ -499,4 +517,123 @@ report::sc_html_stream& covenant_state_t::generate_report( report::sc_html_strea
 
   return root;
 }
-} // Namespace covenant ends
+
+void initialize_soulbinds( player_t* player )
+{
+  if ( !player->covenant )
+    return;
+
+  for ( auto soulbind_spell : player->covenant->soulbind_spells() )
+  {
+    auto spell = player->find_spell( soulbind_spell );
+    if ( !spell->ok() )
+      continue;
+
+    special_effect_t effect { player };
+    effect.type = SPECIAL_EFFECT_EQUIP;
+    effect.source = SPECIAL_EFFECT_SOURCE_SOULBIND;
+
+    unique_gear::initialize_special_effect( effect, soulbind_spell );
+
+    player->special_effects.push_back( new special_effect_t( effect ) );
+  }
+}
+
+void register_soulbinds()
+{
+  unique_gear::register_special_effect( 322721, soulbinds::grove_invigoration );
+}
+
+struct covenant_ability_cast_cb_t : public dbc_proc_callback_t
```

Comment:
I think the class here needs to somehow have the "buff" part in the name as well, and the player->covenant->cast_callback variable as well.

It could also make sense to have another class between dbc_proc_callback_t and this class which only adds the part of setting up a convenant ability callback and filtering, without the buff_list in it.

That way it is also easy to set up other covenant ability callbacks which do arbitrary stuff, separated from the players single covenant ability buff triggering callback.

Reply:
I was thinking since procs are either buffs or actions, if we need action procs in the future, we can add an action_list and also iterate over that in the callback trigger, and just keep everything within a single callback.

Can also set up a base class for the elements of the list vector and have subclasses do a trigger() or schedule_execute() or other as appropriate when the list is iterated on.

Reply:
It certainly.simplifies the usage. You could also add both buff and action callbacks into the same callback.